Kashoo/synctos
# Issue 338: Deprecated mustNotBeMissing and mustNotBeNull constraints

# Description

The `mustNotBeMissing` and `mustNotBeNull` constraints do not behave as expected since, when Sync Gateway encounters `null` values from a JSON document into an object to pass to the sync function, it removes the properties altogether. The document validation script (`validate-document-definitions`) now fails if it encounters either constraint in document definitions and prints a deprecation warning. Furthermore, the `mustNotBeNull` constraint now fails if the corresponding property value is missing/`undefined` (since `null` values become `undefined` when converted to the object that's passed to a sync function).

# Testing

1. Create a document definition with `mustNotBeMissing` and `mustNotBeNull` constraints:

```javascript
{
  testDoc: {
    typeFilter: simpleTypeFilter,
    channels: { write: 'write' },
    propertyValidators: {
      mustNotBeMissingProp: {
        type: 'string',
        mustNotBeMissing: true
      },
      mustNotBeNullProp: {
        type: 'integer',
        mustNotBeNull: true
      },
      requiredProp: {
        type: 'float',
        required: true
      }
    }
  }
}
```

2. Run the validation script on the test document definitions (it should produce two validation errors indicating that the `mustNotBeMissing` and `mustNotBeNull` constraints are deprecated):

```bash
./validate-document-definitions /path/to/test-doc-definitions.js
```

3. Generate a sync function from the test document definitions:

```bash
./make-sync-function -j build/test-doc-definitions.js build/test-sync-function.txt
```

4. Copy and paste the generated sync function into a Sync Gateway configuration file.
5. Attempt to create a document where `mustNotBeMissingProp` is `null` (it should be rejected):

```
PUT /test/required-doc-1 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"requiredProp": -12.5,
	"mustNotBeMissingProp": null,
	"mustNotBeNullProp": 579
}
```

6. Attempt to create a document where `mustNotBeMissingProp` is missing (it should be rejected):

```
PUT /test/required-doc-2 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"requiredProp": -12.5,
	"mustNotBeNullProp": 579
}
```

7. Attempt to create a document where `mustNotBeNullProp` is `null` (it should be rejected):

```
PUT /test/required-doc-3 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"requiredProp": -12.5,
	"mustNotBeMissingProp": "foobar",
	"mustNotBeNullProp": null
}
```

8. Attempt to create a document where `mustNotBeNullProp` is missing (it should be rejected):

```
PUT /test/required-doc-4 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"requiredProp": -12.5,
	"mustNotBeMissingProp": "foobar"
}
```

9. Attempt to create a document where `requiredProp` is `null` (it should be rejected):

```
PUT /test/required-doc-5 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"requiredProp": null,
	"mustNotBeMissingProp": "foobar",
	"mustNotBeNullProp": 579
}
```

10. Attempt to create a document where `requiredProp` is missing (it should be rejected):

```
PUT /test/required-doc-6 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"mustNotBeMissingProp": "foobar",
	"mustNotBeNullProp": 579
}
```

11. Create a document where all properties are supplied (it should be accepted):

```
PUT /test/required-doc-7 HTTP/1.1
Host: localhost:4985
Content-Type: application/json
{
	"type": "testDoc",
	"requiredProp": -12.5,
	"mustNotBeMissingProp": "foobar",
	"mustNotBeNullProp": 579
}
```

# Related Issue

* GitHub issue link: #338
